### ISSUE TITLE
Using Hard Negtive NCE For Training

### ISSUE THREAD

**Issue by VincentVanNF** [TIGER-AI-Lab/VLM2Vec#37]
Based on the hard negative loss function mentioned in the reference paper [Filtering, Distillation, and Hard Negatives for Vision-Language Pre-Training](https://arxiv.org/abs/2301.02280)

![Image](https://github.com/user-attachments/assets/3d7a6910-ae13-4cd7-b8e7-a303a2c5225b)

I implemented an experiment based on this function in VLM2Vec, and the results of my downstream task experiments show that hard negative sampling performs better than the ordinary NCE loss. Below is the hard negative loss function I implemented in the batch, and I hope it can be adopted.

```
class MMEBModel(nn.Module):
....
  def batch_nce_loss(self, q_reps:torch.Tensor, t_reps:torch.Tensor, labels:torch.Tensor,temperature,hard_nce:bool=False):
          if( not hard_nce):
              self.hdnce_alpha = 1.0
              self.hdnce_beta = 0.0
          
          # batch_size = q_reps.size(0)
          
          scores = self.compute_similarity(q_reps, t_reps)
          scores = scores.view(q_reps.size(0), -1)
          # scores = scores / self.temperature
          # print("temperature:{}".format(temperature))
          scores = scores / temperature
          # target = torch.arange(scores.size(0), device=scores.device, dtype=torch.long)
          # target = target * (q_reps.size(0) // t_reps.size(0))
          
          # test&debug: labels  = torch.tensor([1,2,3,1],device=labels.device)
          # rank0_print(scores)
          '''
              for ideal condition:all labels are different
              mask = [
                  [1.0, 0.0, 0.0, 0.0],
                  [0.0, 1.0, 0.0, 0.0],
                  [0.0, 0.0, 1.0, 0.0]
                  [0.0, 0.0, 0.0, 1.0]
              ]
          '''
          mask = torch.eq(labels.unsqueeze(1), labels.unsqueeze(1).T).float()
  
          # rank0_print("scores:{}".format(scores))
          # rank0_print("mask:{}".format(mask))
          pos_scores = scores.masked_fill(~mask.bool(), float('-inf'))
          sum_exp_pos_scores = torch.sum(torch.exp(pos_scores), dim=-1)
          # rank0_print("pos_scores:{}".format(pos_scores))
          # rank0_print("sum_exp_pos_scores:{}".format(sum_exp_pos_scores))
  
          neg_scores = scores.masked_fill(mask.bool(), float('-inf'))
          beta_neg_scores = scores * self.hdnce_beta
          # beta_neg_scores = scores * 0.0002
          beta_neg_scores = beta_neg_scores.masked_fill(mask.bool(), float('-inf'))
          sum_of_mask_per_row = (1-mask).sum(dim=-1).unsqueeze(-1).to(dtype=beta_neg_scores.dtype)
          neg_weight = torch.softmax(beta_neg_scores, dim=-1) *  sum_of_mask_per_row
          # rank0_print("neg_scores:{}".format(neg_scores))
          # rank0_print("neg_weight:{}".format(neg_weight))
  
          exp_neg_scores = torch.exp(neg_scores)
          weighted_neg_scores = exp_neg_scores * neg_weight
  
          epsilon = 1e-8
          log_sum_exp_pos = torch.log(sum_exp_pos_scores + epsilon)
          sum_exp_all = self.hdnce_alpha * sum_exp_pos_scores + torch.sum(weighted_neg_scores, dim=-1)
          # rank0_print("sum_exp_all:{}".format(sum_exp_all))
          # rank0_print("sum_exp_pos_scores:{}".format(sum_exp_pos_scores))
  
          log_sum_exp_all = torch.log(sum_exp_all + epsilon)
          
          # log_sum_exp_pos = torch.logsumexp(pos_scores, dim=-1)
          # log_sum_exp_all = torch.logsumexp(scores, dim=-1)
  
          loss = (log_sum_exp_all - log_sum_exp_pos).mean()
          # loss = self.cross_entropy(scores / self.temperature, target)
          # exit()
          if self.is_ddp:
              loss = loss * self.world_size
          return loss
```

**Comment by memray** [comment]
Hi @VincentVanNF , 

Much appreciated for the recommendation! We are investigating hard negatives and will definitely look into this method. And feel free to create a PR to incorporate your code!

Thank you!
Rui
